File: rules-interpreter/src/main/java/com/github/iamrenny/ruleflow/evaluators/UnaryContextEvaluator.java
package com.github.iamrenny.ruleflow.evaluators;

import com.github.iamrenny.ruleflow.RuleFlowLanguageLexer;
import com.github.iamrenny.ruleflow.RuleFlowLanguageParser.UnaryContext;
import com.github.iamrenny.ruleflow.visitors.Visitor;

public class UnaryContextEvaluator implements ContextEvaluator<UnaryContext> {

    @Override
    public Object evaluate(UnaryContext ctx, Visitor visitor) {
        Object valLeft = visitor.visit(ctx.left);
        Double left = toDoubleOrNull(valLeft.toString());

        if (left == null) {
            throw new IllegalArgumentException("Parameter value not supported: " + valLeft);
        }

        switch (ctx.op.getType()) {
            case RuleFlowLanguageLexer.ABS:
                return Math.abs(left);
            default:
                throw new IllegalArgumentException("Operation not supported: " + ctx.op.getText());
        }
    }

    private Double toDoubleOrNull(String value) {
        try {
            return Double.parseDouble(value);
        } catch (NumberFormatException e) {
            return null;
        }
    }
}

File: analang-interpreter/src/main/kotlin/com/rappi/fraud/rules/parser/evaluators/UnaryContextEvaluator.kt
package com.rappi.fraud.rules.parser.evaluators

import com.rappi.fraud.analang.ANALexer
import com.rappi.fraud.analang.ANAParser
import com.rappi.fraud.rules.parser.visitors.Visitor
import kotlin.math.absoluteValue

class UnaryContextEvaluator : ContextEvaluator<ANAParser.UnaryContext> {

    override fun evaluate(ctx: ANAParser.UnaryContext, visitor: Visitor): Any {
        val valLeft = visitor.visit(ctx.left)
        val left = valLeft.toString().toDoubleOrNull()

        if(left == null)
            throw IllegalArgumentException("parameter value not supported: $valLeft")

        return when (ctx.op.type) {
            ANALexer.ABS -> left.absoluteValue
            else -> throw IllegalArgumentException("Operation not supported: ${ctx.op.text}")
        }

    }

}